File: python/0530.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# class Solution:
# def traverse(self, node:ListNode):
#     if node is None:
#         return float('inf'), -float('inf'), float('inf')

#     al, bl, rl = self.traverse(node.left)
#     ar, br, rr = self.traverse(node.right)

#     a = min(node.val, al, ar)
#     b = max(node.val, bl, br)
#     r = min(abs(node.val-bl), abs(node.val-ar), rl, rr)

#     return a, b, r

# def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
#     a, b, r = self.traverse(root)

#     return r
from itertools import pairwise


class Solution:
    def traverse(self, node: ListNode):
        if node is None:
            return
        self.traverse(node.left)
        self.arr.append(node.val)
        self.traverse(node.right)

    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.arr = []
        self.traverse(root)
        print(self.arr)
        ans = float("inf")
        for a, b in pairwise(self.arr):
            ans = min(ans, b - a)

        return ans

File: c++/0530.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
   public:
    void inorder(TreeNode* node, vector<int>& arr) {
        if (node == nullptr) return;
        inorder(node->left, arr);
        arr.push_back(node->val);
        inorder(node->right, arr);
    }
    int getMinimumDifference(TreeNode* root) {
        auto arr = vector<int>();

        inorder(root, arr);

        int ans = INT_MAX;
        for (int i = 0; i < arr.size() - 1; i++) {
            ans = min(ans, arr[i + 1] - arr[i]);
        }

        return ans;
    }
};
